The drainer may fail when clustered index table is used in the upstream

## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.

The support of clustered index for tidb-binlog https://github.com/pingcap/tidb/pull/24135, there are some cases which are handled incorrectly and will make the drainer fail.

```
CREATE TABLE `t` (
  `col_0` timestamp NOT NULL,
  `col_1` datetime NOT NULL,
  `col_2` int(11) NOT NULL,
  `col_3` int(11) NOT NULL,
  `col_4` varchar(147) NOT NULL,
  `col_5` tinyint(4) DEFAULT NULL,
  `col_6` int(11) NOT NULL,
  `col_7` char(112) DEFAULT NULL,
  `col_8` char(85) DEFAULT NULL,
  `col_9` datetime DEFAULT NULL,
  `col_10` char(242) DEFAULT NULL,
  `col_11` char(210) NOT NULL,
  `col_12` int(11) DEFAULT NULL,
  `col_13` char(125) DEFAULT NULL,
  `col_14` char(135) DEFAULT NULL,
  `col_15` int(11) DEFAULT NULL,
  `col_16` tinyint(4) DEFAULT NULL,
  `col_17` timestamp NULL DEFAULT NULL,
  `col_18` date DEFAULT NULL,
  `col_19` tinyint(4) NOT NULL,
  `col_20` char(97) DEFAULT NULL,
  `col_21` date NOT NULL,
  PRIMARY KEY (`col_0`,`col_1`,`col_2`) /*T![clustered_index] CLUSTERED */,
  KEY `k5` (`col_20`,`col_15`),
  KEY `k6` (`col_19`),
  KEY `k7` (`col_3`,`col_15`,`col_0`,`col_6`),
  KEY `k8` (`col_16`,`col_17`,`col_13`),
  KEY `k9` (`col_17`,`col_2`,`col_13`,`col_6`)
)
```

Create a clustered index in the upstream, start the binlog sync to a down stream mysql instance.
The insert statement in the upstream:

[INFO] [session.go:2854] [GENERAL_LOG] [conn=35] [user=root@127.0.0.1] [schemaVersion=33] [txnStartTS=424839918842282024] [forUpdateTS=424839918842282024] [isReadConsistency=false] [current_db=test] [txn_mode=PESSIMISTIC] [sql="INSERT INTO t VALUES(\"1996-11-13 10:38:39\", \"9472-11-22 10:36:06\", 1709134011, -496258536, \"condescending_colden2\", -89, 1208262486, \"tender_minsky9\", \"kickass_montalcini2\", \"7210-02-19 11:27:27\", \"musing_jang2\", \"awesome_fermat1\", 942386696, \"trusting_brahmagupta3\", \"quizzical_kirch8\", 725385470, 14, \"1993-07-03 03:20:22\", \"6743-06-01\", -81, \"inspiring_aryabhata9\", \"6901-10-14\"), (\"2002-05-29 08:44:47\", \"5010-12-15 05:31:09\", 184748039, 2072274414, \"wonderful_saha9\", 7, -1268825709, \"pedantic_murdock9\", \"competent_stonebraker3\", \"7751-03-16 23:22:48\", \"confident_gates0\", \"kickass_goodall7\", -644542979, \"competent_mestorf7\", \"sad_ptolemy2\", 1985148604, -7, \"1993-08-29 01:29:08\", \"4009-04-11\", -116, \"competent_mclean8\", \"5107-10-05\"), (\"2009-06-29 21:55:30\", \"6952-05-24 09:53:10\", -408961458, 352417131, \"modest_jennings5\", 96, 883422748, \"sad_chandrasekhar2\", \"goofy_keller8\", \"7827-11-16 17:20:36\", \"nervous_curran9\", \"optimistic_minsky6\", 1554766506, \"hungry_minsky1\", \"epic_colden3\", 2053703972, -6, \"1979-02-23 00:06:16\", \"5738-05-03\", 114, \"determined_turing9\", \"5522-07-14\"), (\"1985-06-03 03:39:45\", \"7821-03-06 13:51:48\", 1547027164, 2122770155, \"sleepy_payne1\", 109, 742587698, \"eloquent_shirley3\", \"sharp_dijkstra6\", \"4139-07-15 04:43:31\", \"zealous_lumiere9\", \"fervent_morse6\", -857810148, \"distracted_banach6\", \"practical_mahavira5\", -363544408, 111, \"1984-09-19 04:36:23\", \"4605-03-22\", 90, \"laughing_bell4\", \"9828-10-22\"), (\"1979-07-28 20:18:45\", \"4835-09-07 09:11:46\", -694237080, 25201435, \"focused_galileo0\", 115, 746918708, \"nifty_swartz4\", \"eager_haibt2\", \"8748-08-12 16:25:38\", \"optimistic_lamarr8\", \"stupefied_ramanujan5\", 1954196882, \"musing_wright5\", \"suspicious_franklin9\", -698179720, 51, \"2007-04-02 22:10:44\", \"4618-07-28\", 88, \"nifty_lichterman8\", \"3368-03-31\"), (\"2019-08-07 17:58:27\", \"6671-05-02 18:01:11\", -1446605472, -2027143714, \"quirky_cray2\", -14, 1417281299, \"frosty_curran0\", \"frosty_davinci9\", \"8352-08-08 03:11:26\", \"inspiring_leavitt0\", \"kind_wilson3\", -880601604, \"wizardly_galileo0\", \"stupefied_morse7\", 1009775090, 28, \"1998-01-08 02:37:58\", \"8973-07-29\", -71, \"nervous_carson7\", \"6045-11-24\"), (\"2019-12-18 21:21:12\", \"7137-03-13 09:33:13\", 166843807, -1154471117, \"distracted_tesla1\", 37, -1154931902, \"brave_minsky6\", \"unruffled_bell8\", \"1021-08-22 10:26:51\", \"zealous_einstein6\", \"focused_northcutt4\", 1521678417, \"inspiring_hodgkin5\", \"determined_blackwell9\", 852704539, -33, \"1986-02-20 06:03:35\", \"4301-02-06\", -65, \"zealous_goldwasser4\", \"6863-04-24\"), (\"1974-12-09 19:54:06\", \"1855-10-29 21:37:36\", -1504221740, 137764080, \"condescending_bose1\", 27, 2140611339, \"stupefied_almeida1\", \"compassionate_newton0\", \"8706-09-22 11:47:37\", \"adoring_liskov2\", \"gifted_shannon2\", 1467456394, \"dazzling_brahmagupta5\", \"distracted_bhabha0\", 1587945068, 65, \"1988-08-21 02:22:55\", \"4016-04-27\", 29, \"zealous_chandrasekhar6\", \"4715-02-12\"), (\"1996-11-22 15:22:36\", \"3878-01-29 11:33:06\", -1744618073, 300131076, \"jolly_noether5\", 73, -562535852, \"objective_austin0\", \"clever_tesla6\", \"9355-09-15 04:09:29\", \"hopeful_meninsky3\", \"gallant_goldberg9\", 518643906, \"upbeat_hoover5\", \"cocky_shannon0\", 1035348622, 46, \"1981-12-17 06:42:14\", \"6021-11-11\", 7, \"nervous_blackwell9\", \"2154-07-07\"), (\"1980-09-15 06:26:32\", \"9765-08-26 22:31:07\", 1778816643, -1256925771, \"zealous_fermi2\", -112, -1514721287, \"elegant_cori0\", \"awesome_bhabha9\", \"2843-01-10 02:27:37\", \"wonderful_jepsen5\", \"elastic_ptolemy2\", 1275670419, \"optimistic_bhabha4\", \"competent_lichterman4\", 1000519870, -125, \"2022-07-07 21:45:42\", \"4623-05-31\", 62, \"epic_jennings4\", \"8654-03-10\")"]


2. What did you expect to see?
The binlog works normally.


3. What did you see instead?
The drainer fail with error logs

[ERROR] [executor.go:111] ["Exec fail, will rollback"] [query="REPLACE INTO `test`.`t`(`col_0`,`col_1`,`col_10`,`col_11`,`col_12`,`col_13`,`col_14`,`col_15`,`col_16`,`col_17`,`col_18`,`col_19`,`col_2`,`col_20`,`col_21`,`col_3`,`col_4`,`col_5`,`col_6`,`col_7`,`col_8`,`col_9`) VALUES(?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)"] [args="[\"1806667743223611392\",\"1697711208401993728\",\"adoring_liskov2\",\"gifted_shannon2\",1467456394,\"dazzling_brahmagupta5\",\"distracted_bhabha0\",1587945068,65,\"1988-08-21 02:22:55\",\"4016-04-27\",29,-1504221740,\"zealous_chandrasekhar6\",\"4715-02-12\",137764080,\"condescending_bose1\",27,2140611339,\"stupefied_almeida1\",\"compassionate_newton0\",\"8706-09-22 11:47:37\"]"] [error="**Error 1292: Incorrect datetime value: '1806667743223611392' for column 'col_0' at row 1**"]



4. Please provide the relate downstream type and version of drainer.
(run `drainer -V` in terminal to get drainer's version)
both nightly branch of 05-09

--- a/drainer/translator/translator.go
+++ b/drainer/translator/translator.go
@@ -74,6 +74,13 @@ func insertRowToDatums(table *model.TableInfo, row []byte) (datums map[int64]typ
 		if err != nil {
 			return nil, errors.Trace(err)
 		}
+		if table.IsCommonHandle {
+			// clustered index could be complex type that need Unflatten from raw datum.
+			aPK, err = tablecodec.Unflatten(aPK, &table.Columns[commonPKInfo.Columns[i].Offset].FieldType, time.Local)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+		}
 		pk = append(pk, aPK)
 	}
 

--- a/tests/dailytest/case.go
+++ b/tests/dailytest/case.go
@@ -43,6 +43,8 @@ var caseClusteredIndexInsert = []string{
 	"create table cluster_t4(a varchar(100), b varchar(100), d varchar(100), c varchar(100), primary key(c(1))) collate utf8mb4_general_ci",
 	"create table cluster_t5(a varchar(100), b varchar(100), d varchar(100), c varchar(100), primary key(c, b(1))) collate utf8mb4_general_ci",
 	"create table cluster_t6(a varchar(100), b varchar(100), d varchar(100), c varchar(100), primary key(c(1), a(1), b)) collate utf8mb4_general_ci",
+	"create table cluster_t7(c1 timestamp, c2 datetime, c3 int, primary key(c1, c2, c3) clustered)",
+	"create table cluster_t8(c1 float, c2 enum('a', 'b'), c3 bit, c4 SET('a', 'b'), c5 time, primary key(c1, c2, c3, c4, c5) clustered)",
 
 	"insert into cluster_t1(a, b, d, c) values ('aaa', 'bbb', 'ddd', 'ccc'), ('111', '222', '444', '333')",
 	"insert into cluster_t2(a, b, d, c) values ('aaa', 'bbb', 'ddd', 'ccc'), ('111', '222', '444', '333')",
@@ -50,6 +52,8 @@ var caseClusteredIndexInsert = []string{
 	"insert into cluster_t4(a, b, d, c) values ('aaa', 'bbb', 'ddd', 'ccc'), ('111', '222', '444', '333')",
 	"insert into cluster_t5(a, b, d, c) values ('aaa', 'bbb', 'ddd', 'ccc'), ('111', '222', '444', '333')",
 	"insert into cluster_t6(a, b, d, c) values ('aaa', 'bbb', 'ddd', 'ccc'), ('111', '222', '444', '333')",
+	"insert into cluster_t7(c1, c2, c3) values ('1996-11-13 10:38:39', '9472-11-22 10:36:06', 1709134011)",
+	"insert into cluster_t8(c1, c2, c3, c4, c5) values (1.1, 'a', 1, 'b', '01:01:01')",
 }
 
 var caseClusteredIndexUpdateNoPK = []string{

